lib/logstash/pipeline.rb

```diff
@@ -183,14 +188,23 @@ def inputworker(plugin)
                              :plugin => plugin.inspect, :error => e))
       end
       puts e.backtrace if @logger.debug?
-      plugin.teardown
+      # input teardown must be synchronized since is can be called concurrently by
+      # the input worker thread and from the pipeline thread shutdown method.
+      # this means that input teardown methods must support multiple calls.
+      @run_mutex.synchronize{plugin.teardown}
```

Comment:
This comment is more an open question than an improvement, so please feel free to educate me if this is too stupid to phrase it here.

I understand this is a feasible way of doing this isolation work in here for the teardown method, cause this code we control not like the plugins, so we make sure this part is separated from the other. 

But I'm not sure if we're over synchronizing here, is this the best place to write use this thread mutex? or can we do it in a more detailed view, might be within the teardown method. Please don't take this as a critic, I like the way is done here, I just have this open question in me :-)